samza-core/src/main/java/org/apache/samza/zk/ZkBarrierForVersionUpgrade.java

```diff
@@ -0,0 +1,166 @@
+    String barrierDonePath = String.format("%s/barrier_done", barrierPath);
+    String barrierProcessors = String.format("%s/barrier_processors", barrierPath);
+
+    zkUtils.makeSurePersistentPathsExists(new String[]{barrierPrefix, barrierPath, barrierProcessors, barrierDonePath});
+
+    // callback for when the barrier is reached
+    Runnable callback = new Runnable() {
+      @Override
+      public void run() {
+        LOG.info("Writing BARRIER DONE to " + barrierDonePath);
+        zkUtils.getZkClient().writeData(barrierDonePath, BARRIER_DONE);
+      }
+    };
+    // subscribe for processor's list changes
+    LOG.info("Subscribing for child changes at " + barrierProcessors);
+    zkUtils.getZkClient().subscribeChildChanges(barrierProcessors,
+        new ZkBarrierChangeHandler(callback, processorsNames));
+  }
+
+  @Override
+  public void waitForBarrier(String version, String processorsName, Runnable callback) {
+    // if participant makes this call it means it has already stopped the old container and got the new job model.
+    String barrierPath = String.format("%s/barrier_%s", barrierPrefix, version);
+    String barrierDonePath = String.format("%s/barrier_done", barrierPath);
+    String barrierProcessors = String.format("%s/barrier_processors", barrierPath);
+    String barrierProcessorThis = String.format("%s/%s", barrierProcessors, processorsName);
+
+
+    // update the barrier for this processor
+    LOG.info("Creating a child for barrier at " + barrierProcessorThis);
+    zkUtils.getZkClient().createPersistent(barrierProcessorThis);
+
+    // now subscribe for the barrier
+    zkUtils.getZkClient().subscribeDataChanges(barrierDonePath, new ZkBarrierReachedHandler(barrierDonePath, debounceTimer, callback));
+  }
+
+  /**
+   * listener for the subscription.
+   */
+  class ZkBarrierChangeHandler implements IZkChildListener {
+    Runnable callback;
+    List<String> names;
+
+    public ZkBarrierChangeHandler(Runnable callback, List<String> names) {
+      this.callback = callback;
+      this.names = names;
+    }
+
+    @Override
+    public void handleChildChange(String parentPath, List<String> currentChildren) throws Exception {
+      // Find out the event & Log
+      boolean allIn = true;
+
+      if (currentChildren == null) {
+        LOG.info("Got handleChildChange with null currentChildren");
+        return;
+      }
+      // debug
+      StringBuilder sb = new StringBuilder();
+      for (String child : currentChildren) {
+        sb.append(child).append(",");
+      }
+      LOG.info("list of children in the barrier = " + parentPath + ":" + sb.toString());
+      sb = new StringBuilder();
+      for (String child : names) {
+        sb.append(child).append(",");
+      }
+      LOG.info("list of children to compare against = " + parentPath + ":" + sb.toString());
+
+
+      // check if all the names are in
+      for (String n : names) {
```

Comment:
I think we definitely discussed this before. Please remind me and perhaps, add some documentation. Do we have to compare the names of the registered nodes necessarily? Isn't the size sufficient?